journalbeat/checkpoint/checkpoint.go

```diff
@@ -0,0 +1,290 @@
+// channel and persists it when the number of changes reaches maxEvents or
+// the amount of time since the last disk write reaches flushInterval.
+func (c *Checkpoint) run() {
+	defer c.wg.Done()
+	defer c.persist()
+
+	flushTimer := time.NewTimer(c.flushInterval)
+	defer flushTimer.Stop()
+loop:
+	for {
+		select {
+		case <-c.done:
+			break loop
+		case s := <-c.save:
+			c.lock.Lock()
+			c.states[s.Path] = s
+			c.lock.Unlock()
+			c.numUpdates++
+			if c.numUpdates < c.maxUpdates {
+				continue
+			}
+		case <-flushTimer.C:
+		}
+
+		c.persist()
+		flushTimer.Reset(c.flushInterval)
+	}
+}
+
+// Shutdown stops the checkpoint worker (which persists any state to disk as
+// it stops). This method blocks until the checkpoint worker shutdowns. Calling
+// this method more once is safe and has no effect.
+func (c *Checkpoint) Shutdown() {
+	c.once.Do(func() {
+		close(c.done)
+		c.wg.Wait()
+	})
+}
+
+// States returns the current in-memory event log state. This state information
+// is bootstrapped with any data found on disk at creation time.
+func (c *Checkpoint) States() map[string]JournalState {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	copy := make(map[string]JournalState)
+	for k, v := range c.states {
+		copy[k] = v
+	}
+
+	return copy
+}
+
+// Persist queues the given event log state information to be written to disk.
+func (c *Checkpoint) Persist(path, cursor string, realTs, monotonicTs uint64) {
+	c.PersistState(JournalState{
+		Path:               path,
+		Cursor:             cursor,
+		RealtimeTimestamp:  realTs,
+		MonotonicTimestamp: monotonicTs,
+	})
+}
+
+// PersistState queues the given event log state to be written to disk.
+func (c *Checkpoint) PersistState(st JournalState) {
+	c.save <- st
+}
+
+// persist writes the current state to disk if the in-memory state is dirty.
+func (c *Checkpoint) persist() bool {
+	if c.numUpdates == 0 {
+		return false
+	}
+
+	err := c.flush()
+	if err != nil {
+		return false
+	}
+
+	logp.Debug("checkpoint", "Checkpoint saved to disk. numUpdates=%d",
+		c.numUpdates)
+	c.numUpdates = 0
+	return true
+}
+
+// flush writes the current state to disk.
+func (c *Checkpoint) flush() error {
+	c.fileLock.Lock()
+	defer c.fileLock.Unlock()
+
+	tempFile := c.file + ".new"
+	file, err := create(tempFile)
+	if os.IsNotExist(err) {
+		// Try to create directory if it does not exist.
+		if createDirErr := c.createDir(); createDirErr == nil {
+			file, err = create(tempFile)
+		}
+	}
+
+	if err != nil {
+		return fmt.Errorf("Failed to flush state to disk. %v", err)
+	}
+
+	// Sort persisted eventLogs by name.
+	c.sort = c.sort[:0]
+	for k := range c.states {
+		c.sort = append(c.sort, k)
+	}
+	sort.Strings(c.sort)
+
+	ps := PersistedState{
+		UpdateTime: time.Now().UTC(),
+		States:     make([]JournalState, len(c.sort)),
+	}
+	for i, name := range c.sort {
+		ps.States[i] = c.states[name]
+	}
+
+	data, err := yaml.Marshal(ps)
+	if err != nil {
+		file.Close()
```

Comment:
Could we use `defer` instead to close the file? Like this it only shows up in one place.